### ISSUE TITLE
W/System.err error on calls to getThumbDrawable on Android API level < 21

### ISSUE THREAD

**Issue by sserdyuk** [NativeScript/NativeScript#4519]
**This is the second time I am reporting this bug. The first report #4471 got closed without proper consideration.** 

I am getting a cryptic error from the runtime, even though everything works as expected. The device is on API 19. The same test runs fine on API 23 and 25, so I am assuming this is for API levels < 21. 

I am using a widget from the compat v7 library to get a better looking "material" widget on the screen. The library has versions and compileSdkVersion API levels, which might be confusing people into thinking that it requires these API levels at run time; it does not. NS currently targets compileSdkVersion 25 and     buildToolsVersion "25.0.2", so everything should work.  

I have made a simple native Android test project, and these methods work fine without any runtime issues.

I have reduced this from a real application to a simple Placeholder component just to demonstrate the problem on a smallest possible setup. Note that even the error is shown in the logs, the function does return the drawable, and I can color it. I get the same message when I call getTrackDrawable(), but not when I call getTextOn(). 

The test app is available at https://github.com/sserdyuk/ns-bug-4471
The native proof-of-concept app is at https://github.com/sserdyuk/ns-bug-4471-android.git

```
<Page xmlns="http://schemas.nativescript.org/tns.xsd" navigatingTo="navigatingTo" class="page">
    <Page.actionBar>
        <ActionBar title="My App" icon="" class="action-bar">
        </ActionBar>
    </Page.actionBar>
    <StackLayout>
        <Placeholder loaded="loaded" creatingView="creatingView"/>
    </StackLayout>
</Page>

```

```
import { EventData } from 'data/observable';
import { Page } from 'ui/page';
import { Color } from 'color';

import { CreateViewEventData } from "ui/placeholder";

export function creatingView(args: CreateViewEventData) {
    console.log("Creating view");
    const nativeView = new android.support.v7.widget.SwitchCompat(args.context);
    // const nativeView = new android.widget.Switch(args.context);
    args.view = nativeView;
}

export function loaded(args) {
    console.log("Loaded");
    let color = new Color("#FF0000");
    args.object.nativeView.getThumbDrawable().setColorFilter(color.android, android.graphics.PorterDuff.Mode.SRC_IN);
}
```

```
Successfully synced application org.nativescript.ThumbDrawable on device LGVS450PP3a1a880.
JS: Creating view
JS: Loaded
W/System.err( 9576): 	at com.tns.Runtime.callJSMethodNative(Native Method)
W/System.err( 9576): 	at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1043)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethodImpl(Runtime.java:925)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethod(Runtime.java:912)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethod(Runtime.java:896)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethod(Runtime.java:888)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethodNative(Native Method)
W/System.err( 9576): 	at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1043)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethodImpl(Runtime.java:925)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethod(Runtime.java:912)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethod(Runtime.java:896)
W/System.err( 9576): 	at com.tns.Runtime.callJSMethod(Runtime.java:888)
```

```
$ tns info
All NativeScript components versions information
┌──────────────────┬─────────────────┬────────────────┬───────────────┐
│ Component        │ Current version │ Latest version │ Information   │
│ nativescript     │ 3.1.1           │ 3.1.1          │ Up to date    │
│ tns-core-modules │ 3.1.0           │ 3.1.0          │ Up to date    │
│ tns-android      │ 3.1.1           │ 3.1.1          │ Up to date    │
│ tns-ios          │                 │ 3.1.0          │ Not installed │
└──────────────────┴─────────────────┴────────────────┴───────────────┘
```

**Comment by NickIliev** [comment]
@sserdyuk I was able to reproduce the issue by creating [the exact same application](https://github.com/NickIliev/NativeScript-Issues-2017/tree/master/NativeScript/issue_4513) as the one you provided (which failed to build due to entry point errors).

The reason for your issue is known native Android inconsistency of the thumb drawable used in the different API levels for SwitchCompat.

e.g.
on API 19
```
console.log(args.object.nativeView.getThumbDrawable());
// android.support.v4.graphics.drawable.DrawableWrapperKitKat@9507bab8
```
on API 21
```
console.log(args.object.nativeView.getThumbDrawable());
// android.graphics.drawable.StateListDrawable@26507f2
```

The difference in the different drawable implementation requires different logic when creating their background drawable and colors. So you will need to do and SDK version check and implement different logic based on the API level.
e.g.
```
if (android.os.Build.VERSION.SDK_INT > 19) {
    console.log(args.object.nativeView.getThumbDrawable());
    args.object.nativeView.getThumbDrawable().setColorFilter(color.android, android.graphics.PorterDuff.Mode.SRC_IN);
} else if (android.os.Build.VERSION.SDK_INT <= 19) {
    // implement the logic for the  KITKAT drawable 
    console.log("on API 19");
    console.log(args.object.nativeView.getThumbDrawable());
}
```

**Comment by sserdyuk** [comment]
@NickIliev Take a look at the project I supplied they build and run. I disagree with your explanation above. In it's simplest form, even simply calling `getThumbDrawable()` without using its result still produces an error. And it should work as demonstrated by the native app example I provided. The issues is somewhere in NS native API bridge, or there's some extra configuration missing. 

Here's the simplified code, still produces an error.

```
export function loaded(args) {
    console.log("Loaded");
    let color = new Color("#FF0000");
    // args.object.nativeView.getThumbDrawable().setColorFilter(color.android, android.graphics.PorterDuff.Mode.SRC_IN);
    args.object.nativeView.getThumbDrawable(); // still produces an error
}

```

**Comment by NickIliev** [comment]
@sserdyuk you are absolutely right - after some additional test it seems that the error is inconsistent and not happening on different environments but I was able to reproduce it on my MAC - we will need some additional time to investigate what is causing this to happen.

**Comment by NickIliev** [comment]
+ 1 reproducible also via SB template 

sample application [here](https://github.com/NickIliev/NativeScript-Issues-2017/tree/master/test/my-app-name)

**Comment by sserdyuk** [comment]
@tsonevn "Backlog" tag doesn't sound encouraging. Isn't ability to call native API the core benefit of using NS? I have a project in beta that's suffering from the issue, and a little more help would be appreciated. Maybe you can give me a workaround for the time being?

**Comment by Pip3r4o** [comment]
I sat down to inspect the problem and find the cause. The complete stack trace can be seen in the device logcat. Here is what mine showed:

```
07-12 04:54:32.113 1861-1861/? W/System.err: java.lang.ClassNotFoundException: android.support.v4.graphics.drawable.DrawableWrapperHoneycomb
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.VMClassLoader.loadClass(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.findClass(ClassLoader.java:761)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.loadClass(ClassLoader.java:821)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.getTypeMetadata(Runtime.java:636)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1043)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:925)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:912)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:896)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:888)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.FragmentClass.onCreateView(FragmentClass.java:35)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Fragment.performCreateView(Fragment.java:1700)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:890)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1062)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.BackStackRecord.run(BackStackRecord.java:684)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1447)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Activity.performResume(Activity.java:5321)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2764)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2803)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2238)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.access$800(ActivityThread.java:135)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Looper.loop(Looper.java:136)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:5001)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invoke(Method.java:515)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
07-12 04:54:32.113 1861-1861/? W/System.err:     at dalvik.system.NativeStart.main(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err: Caused by: java.lang.NoClassDefFoundError: [generic]
07-12 04:54:32.113 1861-1861/? W/System.err: 	... 1 more
07-12 04:54:32.113 1861-1861/? W/System.err: java.lang.ClassNotFoundException: android.support.v4.graphics.drawable.DrawableWrapperGingerbread
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.VMClassLoader.loadClass(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.findClass(ClassLoader.java:761)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.loadClass(ClassLoader.java:821)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.getTypeMetadata(Runtime.java:636)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1043)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:925)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:912)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:896)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:888)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.FragmentClass.onCreateView(FragmentClass.java:35)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Fragment.performCreateView(Fragment.java:1700)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:890)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1062)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.BackStackRecord.run(BackStackRecord.java:684)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1447)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Activity.performResume(Activity.java:5321)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2764)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2803)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2238)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.access$800(ActivityThread.java:135)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Looper.loop(Looper.java:136)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:5001)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invoke(Method.java:515)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
07-12 04:54:32.113 1861-1861/? W/System.err:     at dalvik.system.NativeStart.main(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err: Caused by: java.lang.NoClassDefFoundError: [generic]
07-12 04:54:32.113 1861-1861/? W/System.err: 	... 1 more
07-12 04:54:32.113 1861-1861/? V/JS: android.support.v4.graphics.drawable.DrawableWrapperKitKat@5282aba8
07-12 04:54:32.125 1861-1861/? I/dalvikvm: Could not find method android.content.res.Resources.getDrawable, referenced from method android.support.v7.widget.ResourcesWrapper.getDrawable
07-12 04:54:32.129 1861-1861/? W/dalvikvm: VFY: unable to resolve virtual method 480: Landroid/content/res/Resources;.getDrawable (ILandroid/content/res/Resources$Theme;)Landroid/graphics/drawable/Drawable;
07-12 04:54:32.129 1861-1861/? D/dalvikvm: VFY: replacing opcode 0x6e at 0x0002
07-12 04:54:32.129 1861-1861/? I/dalvikvm: Could not find method android.content.res.Resources.getDrawableForDensity, referenced from method android.support.v7.widget.ResourcesWrapper.getDrawableForDensity
07-12 04:54:32.129 1861-1861/? W/dalvikvm: VFY: unable to resolve virtual method 482: Landroid/content/res/Resources;.getDrawableForDensity (IILandroid/content/res/Resources$Theme;)Landroid/graphics/drawable/Drawable;
07-12 04:54:32.129 1861-1861/? D/dalvikvm: VFY: replacing opcode 0x6e at 0x0002
07-12 04:54:32.129 1861-1861/? E/dalvikvm: Could not find class 'android.graphics.drawable.RippleDrawable', referenced from method android.support.v7.widget.AppCompatImageHelper.hasOverlappingRendering
07-12 04:54:32.129 1861-1861/? W/dalvikvm: VFY: unable to resolve instanceof 157 (Landroid/graphics/drawable/RippleDrawable;) in Landroid/support/v7/widget/AppCompatImageHelper;
07-12 04:54:32.129 1861-1861/? D/dalvikvm: VFY: replacing opcode 0x20 at 0x000c
```

We can now see that 2 classes failed to load, but if they did, wouldn't the application crash?

The responsible party for the above is the Android Runtime - when the DrawableWrapperKitkat class is requested, all its super classes are also sought in the android class loader, and subsequently loaded. If a class is not found in the first class loader, the runtime looks for it in the 2nd one, and so on, and so on, until the class is finally found, or not at all, at which point your app would crash.

And here are the lines doing what I just described - https://github.com/NativeScript/android-runtime/blob/b46cf87b00ade1b8c849e9a8d8dab9ba6e1287c6/runtime/src/main/java/com/tns/Runtime.java#L642

I'll ensure the message is more verbose, explaining what had just happened, while still trying not to confuse the developer

If what you are saying in your original post is correct, you should not be needing a workaround. The core benefit of using NS - the native API access is there, and doesn't appear to spoil the runtime execution of your application. What impact does the printing of a stacktrace have on your project?

And for the sake of completion, here's the TypeScript code which mimics what the Android Studio application you provided does.

_main-page.ts_ (Modified @NickIliev's sample app_)
```TypeScript
import { EventData } from 'data/observable';
import { Page } from 'ui/page';
import { Color } from 'color';

import { CreateViewEventData } from "ui/placeholder";

export function creatingView(args: CreateViewEventData) {
    console.log("Creating view");
    const nativeView = new android.support.v7.widget.SwitchCompat(args.context);
    // const nativeView = new android.widget.Switch(args.context);
    args.view = nativeView;
    args.view.setOnCheckedChangeListener(new android.support.v7.widget.SwitchCompat.OnCheckedChangeListener({
        onCheckedChanged: (view, state) => {
            if (state) {
                (<any>view).getThumbDrawable().setColorFilter(android.graphics.Color.RED, android.graphics.PorterDuff.Mode.SRC_IN);
            } else {
                (<any>view).getThumbDrawable().setColorFilter(android.graphics.Color.BLUE, android.graphics.PorterDuff.Mode.SRC_IN);
            }
        }
    }))
}

export function loaded(args) {
    console.log("Loaded");
    let color = new Color("#FF0000");

    console.log(args.object.nativeView.getThumbDrawable());
    args.object.nativeView.getThumbDrawable().setColorFilter(color.android, android.graphics.PorterDuff.Mode.SRC_IN);
}
```

**Comment by Pip3r4o** [comment]
@sserdyuk 
I sat down to inspect the problem and find the cause. The complete stack trace can be seen in the device logcat. Here is what mine showed:

```
07-12 04:54:32.113 1861-1861/? W/System.err: java.lang.ClassNotFoundException: android.support.v4.graphics.drawable.DrawableWrapperHoneycomb
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.VMClassLoader.loadClass(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.findClass(ClassLoader.java:761)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.loadClass(ClassLoader.java:821)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.getTypeMetadata(Runtime.java:636)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1043)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:925)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:912)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:896)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:888)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.FragmentClass.onCreateView(FragmentClass.java:35)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Fragment.performCreateView(Fragment.java:1700)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:890)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1062)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.BackStackRecord.run(BackStackRecord.java:684)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1447)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Activity.performResume(Activity.java:5321)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2764)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2803)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2238)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.access$800(ActivityThread.java:135)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Looper.loop(Looper.java:136)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:5001)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invoke(Method.java:515)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
07-12 04:54:32.113 1861-1861/? W/System.err:     at dalvik.system.NativeStart.main(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err: Caused by: java.lang.NoClassDefFoundError: [generic]
07-12 04:54:32.113 1861-1861/? W/System.err: 	... 1 more
07-12 04:54:32.113 1861-1861/? W/System.err: java.lang.ClassNotFoundException: android.support.v4.graphics.drawable.DrawableWrapperGingerbread
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.VMClassLoader.loadClass(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.findClass(ClassLoader.java:761)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.BootClassLoader.loadClass(ClassLoader.java:821)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.getTypeMetadata(Runtime.java:636)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1043)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethodImpl(Runtime.java:925)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:912)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:896)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.Runtime.callJSMethod(Runtime.java:888)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.tns.FragmentClass.onCreateView(FragmentClass.java:35)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Fragment.performCreateView(Fragment.java:1700)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:890)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1062)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.BackStackRecord.run(BackStackRecord.java:684)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1447)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.Activity.performResume(Activity.java:5321)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.performResumeActivity(ActivityThread.java:2764)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:2803)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2238)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.access$800(ActivityThread.java:135)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.os.Looper.loop(Looper.java:136)
07-12 04:54:32.113 1861-1861/? W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:5001)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err:     at java.lang.reflect.Method.invoke(Method.java:515)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
07-12 04:54:32.113 1861-1861/? W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
07-12 04:54:32.113 1861-1861/? W/System.err:     at dalvik.system.NativeStart.main(Native Method)
07-12 04:54:32.113 1861-1861/? W/System.err: Caused by: java.lang.NoClassDefFoundError: [generic]
07-12 04:54:32.113 1861-1861/? W/System.err: 	... 1 more
07-12 04:54:32.113 1861-1861/? V/JS: android.support.v4.graphics.drawable.DrawableWrapperKitKat@5282aba8
07-12 04:54:32.125 1861-1861/? I/dalvikvm: Could not find method android.content.res.Resources.getDrawable, referenced from method android.support.v7.widget.ResourcesWrapper.getDrawable
07-12 04:54:32.129 1861-1861/? W/dalvikvm: VFY: unable to resolve virtual method 480: Landroid/content/res/Resources;.getDrawable (ILandroid/content/res/Resources$Theme;)Landroid/graphics/drawable/Drawable;
07-12 04:54:32.129 1861-1861/? D/dalvikvm: VFY: replacing opcode 0x6e at 0x0002
07-12 04:54:32.129 1861-1861/? I/dalvikvm: Could not find method android.content.res.Resources.getDrawableForDensity, referenced from method android.support.v7.widget.ResourcesWrapper.getDrawableForDensity
07-12 04:54:32.129 1861-1861/? W/dalvikvm: VFY: unable to resolve virtual method 482: Landroid/content/res/Resources;.getDrawableForDensity (IILandroid/content/res/Resources$Theme;)Landroid/graphics/drawable/Drawable;
07-12 04:54:32.129 1861-1861/? D/dalvikvm: VFY: replacing opcode 0x6e at 0x0002
07-12 04:54:32.129 1861-1861/? E/dalvikvm: Could not find class 'android.graphics.drawable.RippleDrawable', referenced from method android.support.v7.widget.AppCompatImageHelper.hasOverlappingRendering
07-12 04:54:32.129 1861-1861/? W/dalvikvm: VFY: unable to resolve instanceof 157 (Landroid/graphics/drawable/RippleDrawable;) in Landroid/support/v7/widget/AppCompatImageHelper;
07-12 04:54:32.129 1861-1861/? D/dalvikvm: VFY: replacing opcode 0x20 at 0x000c
```

We can now see that 2 classes failed to load, but if they did, wouldn't the application crash?

The responsible party for the above is the Android Runtime - when the DrawableWrapperKitkat class is requested, all its super classes are also sought in the android class loader, and subsequently loaded. If a class is not found in the first class loader, the runtime looks for it in the 2nd one, and so on, and so on, until the class is finally found, or not at all, at which point your app would crash.

And here are the lines doing what I just described - https://github.com/NativeScript/android-runtime/blob/b46cf87b00ade1b8c849e9a8d8dab9ba6e1287c6/runtime/src/main/java/com/tns/Runtime.java#L642

I'll ensure the message is more verbose, explaining what had just happened, while still trying not to confuse the developer

If what you are saying in your original post is correct, you should not be needing a workaround. The core benefit of using NS - the native API access is there, and doesn't appear to spoil the runtime execution of your application. What impact does the printing of a stacktrace have on your project?

And for the sake of completion, here's the TypeScript code which mimics what the Android Studio application you provided does.

_main-page.ts_ (Modified @NickIliev's sample app_)
```TypeScript
import { EventData } from 'data/observable';
import { Page } from 'ui/page';
import { Color } from 'color';

import { CreateViewEventData } from "ui/placeholder";

export function creatingView(args: CreateViewEventData) {
    console.log("Creating view");
    const nativeView = new android.support.v7.widget.SwitchCompat(args.context);
    // const nativeView = new android.widget.Switch(args.context);
    args.view = nativeView;
    args.view.setOnCheckedChangeListener(new android.support.v7.widget.SwitchCompat.OnCheckedChangeListener({
        onCheckedChanged: (view, state) => {
            if (state) {
                (<any>view).getThumbDrawable().setColorFilter(android.graphics.Color.RED, android.graphics.PorterDuff.Mode.SRC_IN);
            } else {
                (<any>view).getThumbDrawable().setColorFilter(android.graphics.Color.BLUE, android.graphics.PorterDuff.Mode.SRC_IN);
            }
        }
    }))
}

export function loaded(args) {
    console.log("Loaded");
    let color = new Color("#FF0000");

    console.log(args.object.nativeView.getThumbDrawable());
    args.object.nativeView.getThumbDrawable().setColorFilter(color.android, android.graphics.PorterDuff.Mode.SRC_IN);
}
```

**Comment by sserdyuk** [comment]
@Pip3r4o Thank you Peter. If you saw, I did make a NS test project, that was failing with the error. I didn't get the verbose error you posted, only the short one. The effect of it is that I can't style a native element with colors. If I try to assign colors directly, I can an exception, if I use css and properties, it just prints it, and css engine eats the error. 
I am puzzled with this problem, because (1) the compat library widgets are already used in NS core modules without any special code to load anything, (2) the native project I made to mimic the issue has no issue and everything just works. 
The only plausible reason I can think of is that NS project builds with compileSdkVersion less than needed by the compat library for this control, which is 24. NS currently sets compileSdkVersion to 22 unless a parameter is passed. I have already tried passing the parameter and setting compileSdkVersion to 25 in the app.gradle, but nothing made a difference. Could you take a look at that test app and see if something obscure is missing there to get things working? https://github.com/sserdyuk/ns-bug-4471

**Comment by Pip3r4o** [comment]
@sserdyuk according to your original post - "I am getting a cryptic error from the runtime, even though everything works as expected. The device is on API 19." - _everything works as expected_, and there are no crashes, despite the short stacktrace you are seeing. What I am trying to make clear here is that, despite the fact that you are seeing the 'error', it is harmless, and should not disturb your app's execution. 

Assigning colors directly to the drawable instance of the switch as shown in my code sample works as expected, and changes the button color when the switch is toggled, yes, I see the error the first time in the console, however the application continues functioning. Tested on device API 19.

As for your concerns with the compileSdkVersion - the CLI will automatically pick the highest android platform you have downloaded, and will use that one, along with the support libraries of the highest matching major version. (e.g. your highest installed api level is `platform-25`, and your extra support libraries installed from the sdk are of version `25.3.1`, those versions will be passed on to the gradle build script.) Having said that, it should be unnecessary to explicitly set the dependency versions in your `app.gradle`, and the very fact that the application doesn't crash while creating an instance of the `android/support/v7/widget/SwitchCompat` class indicates that your support libraries are of a version high enough that supports it.

What device are you testing the code samples on? I've tested locally on Geny 19, a Nexus 4 (API 19), Android Studio 19 emulator, the message in the console appeared almost consistently for all 3 devices, but the switch button had its color changed.
